python/pyspark/sql/connect/session.py

```diff
@@ -215,7 +215,31 @@ def createDataFrame(
         _inferred_schema: Optional[StructType] = None
 
         if isinstance(data, pd.DataFrame):
-            _table = pa.Table.from_pandas(data)
+            from pandas.api.types import (  # type: ignore[attr-defined]
+                is_datetime64_dtype,
+                is_datetime64tz_dtype,
+            )
+
+            # We need double conversions for the truncation, first truncate to microseconds.
+            for col in data:
+                print("Checking", col)
+                if is_datetime64tz_dtype(data[col].dtype):
+                    data[col] = data[col].astype("datetime64[us, UTC]")
+                elif is_datetime64_dtype(data[col].dtype):
+                    data[col] = data[col].astype("datetime64[us]")
+
+            # Create a new schema and change the types to the truncated microseconds.
+            pd_schema = pa.Schema.from_pandas(data)
+            new_schema = pa.schema([])
+            for x in range(len(pd_schema.types)):
+                f = pd_schema.field(x)
+                if isinstance(f.type, pa.TimestampType) and f.type.unit == "ns":
```

Comment:
what if the `TimestampType` is in a nested types? For example, `Struct[Long, TimestampType]` ?

maybe add a TODO here

Reply:
Done